plat/kvm/efi.c

```diff
@@ -202,12 +201,22 @@ static int uk_efi_md_to_bi_mrd(struct uk_efi_mem_desc *const md,
 	return 0;
 }
 
-static void uk_efi_get_mmap(struct uk_efi_mem_desc **map, uk_efi_uintn_t *map_sz,
-			    uk_efi_uintn_t *desc_sz)
+static void uk_efi_get_mmap(uk_efi_uintn_t *uk_efi_map_key,
+			    struct uk_efi_mem_desc **map,
+			    uk_efi_uintn_t *map_sz, uk_efi_uintn_t *desc_sz)
 {
 	uk_efi_status_t status;
 	__u32 desc_ver;
 
+	/* The MemoryMap buffer may has been allocated by the previous
+	 * invocation
+	 */
+	if (*map != NULL) {
+		status = uk_efi_bs->free_pool(*map);
```

Comment:
Can't we do this retry in the same function still? :thinking:
It's just that this `*map != NULL` check feels unnatural and unsafe. How about we use a locally declared variable . How about we do something like this instead
```C
static void uk_efi_get_mmap(struct uk_efi_mem_desc **map, uk_efi_uintn_t *map_sz,
			    uk_efi_uintn_t *desc_sz)
{
	uk_efi_uintn_t *map_key;
	uk_efi_status_t status;
	__u32 desc_ver;
	__u8 retries;

	retries = 0;
uk_efi_get_mmap_retry:
	if (retries) {
		if (unlikely(retries > 1))
			UK_EFI_CRASH("Failed ExitBootServices a second time\n");

		status = uk_efi_bs->free_pool(*map);
		if (unlikely(status != UK_EFI_SUCCESS))
			UK_EFI_CRASH("Failed to free previous memory map\n");
	}

	/* As the UEFI Spec says:
	 * If the MemoryMap buffer is too small, the EFI_BUFFER_TOO_SMALL
	 * error code is returned and the MemoryMapSize value contains the
	 * size of the buffer needed to contain the current memory map. The
	 * actual size of the buffer allocated for the consequent call to
	 * GetMemoryMap() should be bigger then the value returned in
	 * MemoryMapSize, since allocation of the new buffer may potentially
	 * increase memory map size.
	 */
	*map_sz = 0;  /* force EFI_BUFFER_TOO_SMALL */
	*map = NULL;
	status = uk_efi_bs->get_memory_map(map_sz, *map, &map_key,
					   desc_sz, &desc_ver);
	if (unlikely(status != UK_EFI_BUFFER_TOO_SMALL))
		UK_EFI_CRASH("Failed to call initial dummy get_memory_map\n");

	/* Make sure the actual allocated buffer is bigger */
	*map_sz += *desc_sz * UK_EFI_SURPLUS_MEM_DESC_COUNT;
	status = uk_efi_bs->allocate_pool(UK_EFI_LOADER_DATA, *map_sz,
					  (void **)map);
	if (unlikely(status != UK_EFI_SUCCESS))
		UK_EFI_CRASH("Failed to allocate memory for map\n");

	/* Now we call it for real */
	status = uk_efi_bs->get_memory_map(map_sz, *map, &map_key,
					   desc_sz, &desc_ver);
	if (unlikely(status != UK_EFI_SUCCESS))
		UK_EFI_CRASH("Failed to get memory map\n");

	status = uk_efi_bs->exit_boot_services(uk_efi_sh, map_key);
	if (status != UK_EFI_SUCCESS) {
		uk_efi_pr_debug("Failed ExitBootServices\n");
		retries++;
		goto uk_efi_get_mmap_retry;
	}
}
```
Wrote this mostly as "pseudocode"<sup>TM</sup>, haven't tested it yet

Reply:
I updated `uk_efi_get_mmap`  to follow this. 
 
> Can't we do this retry in the same function still? 🤔
Imo, doing it in the same function works fine in this case, but API design-wise, `uk_efi_get_mmap`   would perform two jobs at a time. The caller might not expect that getting the memory map would exit the Boot Service :smiley_cat:. How about renaming it into `uk_efi_get_mmap_and_exit`?

Reply:
> I updated `uk_efi_get_mmap` to follow this.

Thank you :pray:.

> How about renaming it into `uk_efi_get_mmap_and_exit`?

Good idea, love it :tiger:.